File: packages/backend/src/auth/service.ts
import type {
  AuthChallenge,
  AuthCompleteCertRequest,
  AuthCompleteRequest,
  DevAuthCompleteRequest,
  DevSessionToken,
  SessionToken
} from "../../../shared/src/index.ts";

import { HttpError } from "../http.ts";
import { randomId, randomServerId } from "../ids.ts";
import type { Env } from "../env.ts";
import type { SharedWorldRepository } from "../repository.ts";
import {
  buildCertificateSignedPayload,
  decodeBase64Field,
  verifyCertificateSignature,
  verifyNonceSignature
} from "./certificate.ts";
import { MojangServicesKeyProvider, type ServicesKeyProvider } from "./services-keys.ts";

const PLAYER_NAME_PATTERN = /^\w{1,16}$/;

export class AuthDomainService {
  private readonly servicesKeys: ServicesKeyProvider;

  constructor(
    private readonly repository: SharedWorldRepository,
    private readonly env: Env
  ) {
    this.servicesKeys = new MojangServicesKeyProvider(repository, env);
  }

  async createChallenge(now = new Date()): Promise<AuthChallenge> {
    const challenge = {
      serverId: randomServerId(),
      expiresAt: new Date(now.getTime() + 5 * 60_000).toISOString(),
      usedAt: null
    };
    await this.repository.createChallenge(challenge);
    return {
      serverId: challenge.serverId,
      expiresAt: challenge.expiresAt
    };
  }

  /**
   * The legacy joinServer/hasJoined flow, only reachable by clients ≤0.2.1.
   * Mojang answers 403 to all Cloudflare Workers egress on sessionserver, so
   * this flow can NEVER verify an identity from production — every subrequest
   * it would make is a guaranteed failure. Answer immediately with the update
   * notice instead: no Mojang egress, no D1 reads. Shipped legacy clients
   * render this message verbatim and treat the code as terminal.
   */
  async completeAuth(_request: AuthCompleteRequest): Promise<SessionToken> {
    throw new HttpError(
      403,
      "identity_verification_failed",
      "Minecraft no longer accepts the sign-in method used by SharedWorld 0.2.1 and older. Please update SharedWorld to the latest version."
    );
  }

  /**
   * Certificate-based auth: the client proves account ownership by signing
   * the challenge nonce with its Mojang-certified profile key. Fully offline
   * against the cached services key set — no sessionserver involved (Mojang
   * blocks that endpoint for Cloudflare Workers egress).
   */
  async completeCertAuth(request: AuthCompleteCertRequest, now = new Date()): Promise<SessionToken> {
    try {
      return await this.completeCertAuthChecked(request, now);
    } catch (error) {
      if (error instanceof HttpError) {
        // 4xx auth failures never reach errorResponse's >=500 logging — so
        // this line is the only way production logs can answer "is a real
        // certificate being rejected?".
        console.warn("SharedWorld certificate auth rejected", {
          code: error.code,
          status: error.status,
          playerName: request.playerName,
          playerUuid: request.playerUuid
        });
      }
      throw error;
    }
  }

  private async completeCertAuthChecked(request: AuthCompleteCertRequest, now: Date): Promise<SessionToken> {
    const challenge = await this.repository.getChallenge(request.serverId);
    if (!challenge) {
      throw new HttpError(404, "challenge_not_found", "Authentication challenge not found.");
    }
    if (challenge.usedAt) {
      throw new HttpError(409, "challenge_used", "Authentication challenge has already been used.");
    }
    if (new Date(challenge.expiresAt).getTime() < now.getTime()) {
      throw new HttpError(410, "challenge_expired", "Authentication challenge has expired.");
    }

    const playerUuid = (request.playerUuid ?? "").toLowerCase();
    const playerName = request.playerName ?? "";
    if (!PLAYER_NAME_PATTERN.test(playerName)) {
      throw new HttpError(400, "invalid_player_name", "Player name is not a valid Minecraft name.");
    }
    const publicKeyDer = decodeBase64Field(
      request.publicKey ?? "",
      "certificate_invalid",
      "Minecraft profile certificate is invalid."
    );
    const keySignature = decodeBase64Field(
      request.keySignature ?? "",
      "certificate_invalid",
      "Minecraft profile certificate is invalid."
    );
    const nonceSignature = decodeBase64Field(
      request.nonceSignature ?? "",
      "signature_invalid",
      "Challenge signature is invalid."
    );

    if (!Number.isFinite(request.publicKeyExpiresAtMs) || request.publicKeyExpiresAtMs < now.getTime()) {
      throw new HttpError(
        403,
        "certificate_expired",
        "Your Minecraft profile keys have expired. Restart the game to refresh them and try again."
      );
    }

    // buildCertificateSignedPayload also rejects malformed UUIDs.
    const payload = buildCertificateSignedPayload(playerUuid, request.publicKeyExpiresAtMs, publicKeyDer);
    const servicesKeys = await this.servicesKeys.playerCertificateKeys();
    if (!(await verifyCertificateSignature(payload, keySignature, servicesKeys))) {
      throw new HttpError(
        403,
        "certificate_invalid",
        "Minecraft profile certificate is not validly signed for this player."
      );
    }
    if (!(await verifyNonceSignature(publicKeyDer, request.serverId, nonceSignature))) {
      throw new HttpError(
        403,
        "signature_invalid",
        "Challenge signature does not match the certified profile key."
      );
    }

    const createdAt = now.toISOString();
    const session = this.createSessionToken(playerUuid, playerName, now);
    await this.repository.markChallengeUsed(request.serverId, createdAt);
    await this.repository.upsertUser({
      playerUuid,
      playerName,
      createdAt
    });
    await this.repository.createSession(session);
    return session;
  }

  async completeDevAuth(request: DevAuthCompleteRequest, now = new Date()): Promise<DevSessionToken> {
    if ((this.env.ALLOW_DEV_AUTH ?? "").toLowerCase() !== "true") {
      throw new HttpError(404, "not_found", "Route not found.");
    }
    if (request.secret !== (this.env.DEV_AUTH_SECRET ?? "")) {
      throw new HttpError(403, "invalid_dev_auth", "SharedWorld developer auth secret is invalid.");
    }

    const createdAt = now.toISOString();
    const session = this.createSessionToken(request.playerUuid, request.playerName, now);
    await this.repository.upsertUser({
      playerUuid: request.playerUuid,
      playerName: request.playerName,
      createdAt
    });
    await this.repository.createSession(session);
    return {
      ...session,
      allowInsecureE4mc: (this.env.ALLOW_DEV_INSECURE_E4MC ?? "").toLowerCase() === "true"
    };
  }

  async getSession(token: string) {
    return this.repository.getSession(token);
  }

  private createSessionToken(playerUuid: string, playerName: string, now: Date): SessionToken {
    return {
      token: randomId("session"),
      playerUuid,
      playerName,
      expiresAt: new Date(
        now.getTime() + Number(this.env.SESSION_TTL_HOURS ?? "168") * 60 * 60_000
      ).toISOString()
    };
  }

}

File: packages/server/crates/sw-core/src/auth/service.rs
//! `AuthDomainService` (`auth/service.ts`).

use std::sync::Arc;

use sw_contracts::{
    AuthChallenge, AuthCompleteCertRequest, DevAuthCompleteRequest, DevSessionToken, SessionToken,
};
use sw_db::repo::{AuthChallengeRecord, UserRecord};
use sw_db::Repository;

use super::certificate::*;
use super::services_keys::ServicesKeyProvider;
use crate::config::Config;
use crate::http_error::{HttpError, HttpResult};
use crate::ids::{random_id, random_server_id};
use crate::time::{self, Instant};

pub struct AuthService {
    repo: Repository,
    config: Arc<Config>,
    keys: Arc<ServicesKeyProvider>,
}

fn valid_player_name(name: &str) -> bool {
    // /^\w{1,16}$/ — JS \w = [A-Za-z0-9_]
    (1..=16).contains(&name.len()) && name.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

impl AuthService {
    pub fn new(repo: Repository, config: Arc<Config>) -> Self {
        let keys = ServicesKeyProvider::new(
            repo.clone(),
            config.mojang_player_certificate_keys.clone(),
            config.mojang_keys_self_refresh,
            config.mojang_publickeys_url.clone(),
        );
        Self { repo, config, keys }
    }

    pub fn services_keys(&self) -> Arc<ServicesKeyProvider> {
        self.keys.clone()
    }

    pub async fn create_challenge(&self, now: Instant) -> HttpResult<AuthChallenge> {
        let challenge = AuthChallengeRecord {
            server_id: random_server_id(),
            expires_at: time::plus_ms_iso(now, 5 * 60_000),
            used_at: None,
        };
        self.repo.create_challenge(challenge.clone()).await?;
        Ok(AuthChallenge { server_id: challenge.server_id, expires_at: challenge.expires_at })
    }

    /// Legacy ≤0.2.1 flow: terminal update notice.
    pub fn complete_auth(&self) -> HttpError {
        HttpError::new(
            403,
            "identity_verification_failed",
            "Minecraft no longer accepts the sign-in method used by SharedWorld 0.2.1 and older. Please update SharedWorld to the latest version.",
        )
    }

    pub async fn complete_cert_auth(
        &self,
        request: &AuthCompleteCertRequest,
        now: Instant,
    ) -> HttpResult<SessionToken> {
        match self.complete_cert_auth_checked(request, now).await {
            Ok(s) => Ok(s),
            Err(e) => {
                // The only line production logs have for "is a real certificate being rejected?".
                tracing::warn!(code = e.code, status = e.status, player_name = %request.player_name, player_uuid = %request.player_uuid, "SharedWorld certificate auth rejected");
                Err(e)
            }
        }
    }

    async fn complete_cert_auth_checked(
        &self,
        request: &AuthCompleteCertRequest,
        now: Instant,
    ) -> HttpResult<SessionToken> {
        let challenge = self.repo.get_challenge(&request.server_id).await?.ok_or_else(|| {
            HttpError::new(404, "challenge_not_found", "Authentication challenge not found.")
        })?;
        if challenge.used_at.is_some() {
            return Err(HttpError::new(
                409,
                "challenge_used",
                "Authentication challenge has already been used.",
            ));
        }
        if time::parse_iso(&challenge.expires_at).is_none_or(|t| t < now) {
            return Err(HttpError::new(410, "challenge_expired", "Authentication challenge has expired."));
        }
        let player_uuid = request.player_uuid.to_lowercase();
        let player_name = request.player_name.clone();
        if !valid_player_name(&player_name) {
            return Err(HttpError::new(
                400,
                "invalid_player_name",
                "Player name is not a valid Minecraft name.",
            ));
        }
        let public_key_der = decode_base64_field(
            &request.public_key,
            "certificate_invalid",
            "Minecraft profile certificate is invalid.",
            403,
        )?;
        let key_signature = decode_base64_field(
            &request.key_signature,
            "certificate_invalid",
            "Minecraft profile certificate is invalid.",
            403,
        )?;
        let nonce_signature = decode_base64_field(
            &request.nonce_signature,
            "signature_invalid",
            "Challenge signature is invalid.",
            403,
        )?;
        let expires_ms = request.public_key_expires_at_ms.as_f64().filter(|v| v.is_finite());
        let Some(expires_ms) = expires_ms.filter(|v| *v >= time::to_millis(now) as f64) else {
            return Err(HttpError::new(
                403,
                "certificate_expired",
                "Your Minecraft profile keys have expired. Restart the game to refresh them and try again.",
            ));
        };
        let payload = build_certificate_signed_payload(&player_uuid, expires_ms as i64, &public_key_der)?;
        let services_keys = self.keys.player_certificate_keys().await?;
        if !verify_certificate_signature(&payload, &key_signature, &services_keys) {
            return Err(HttpError::new(
                403,
                "certificate_invalid",
                "Minecraft profile certificate is not validly signed for this player.",
            ));
        }
        if !verify_nonce_signature(&public_key_der, &request.server_id, &nonce_signature) {
            return Err(HttpError::new(
                403,
                "signature_invalid",
                "Challenge signature does not match the certified profile key.",
            ));
        }
        let created_at = time::to_iso(now);
        let session = self.session_token(&player_uuid, &player_name, now);
        self.repo.mark_challenge_used(&request.server_id, &created_at).await?;
        self.repo
            .upsert_user(UserRecord {
                player_uuid: player_uuid.clone(),
                player_name: player_name.clone(),
                created_at,
            })
            .await?;
        self.repo.create_session(session.clone()).await?;
        Ok(session)
    }

    pub async fn complete_dev_auth(
        &self,
        request: &DevAuthCompleteRequest,
        now: Instant,
    ) -> HttpResult<DevSessionToken> {
        if !self.config.allow_dev_auth {
            return Err(HttpError::new(404, "not_found", "Route not found."));
        }
        if request.secret != self.config.dev_auth_secret.clone().unwrap_or_default() {
            return Err(HttpError::new(
                403,
                "invalid_dev_auth",
                "SharedWorld developer auth secret is invalid.",
            ));
        }
        let created_at = time::to_iso(now);
        let session = self.session_token(&request.player_uuid, &request.player_name, now);
        self.repo
            .upsert_user(UserRecord {
                player_uuid: request.player_uuid.clone(),
                player_name: request.player_name.clone(),
                created_at,
            })
            .await?;
        self.repo.create_session(session.clone()).await?;
        Ok(DevSessionToken { session, allow_insecure_e4mc: self.config.allow_dev_insecure_e4mc })
    }

    pub async fn get_session(&self, token: &str) -> HttpResult<Option<SessionToken>> {
        Ok(self.repo.get_session(token).await?)
    }

    fn session_token(&self, player_uuid: &str, player_name: &str, now: Instant) -> SessionToken {
        SessionToken {
            token: random_id("session"),
            player_uuid: player_uuid.to_string(),
            player_name: player_name.to_string(),
            expires_at: time::plus_ms_iso(now, self.config.session_ttl_hours * 60 * 60_000),
        }
    }
}
